boards: arm: pinetime: Fix button sample

caspermeijn opened the pull request:
The button sample application doesn't work for the pinetime board. This is because the `key_out` gpio must be set high before the `key_in` gpio can be read. 

This adds a new DTS definition for the `key_out` and some board specific code to enable the gpio during boot. The board-specfic sample is removed.

caspermeijn committed:
samples: boards: pinetime: Remove duplicate sample

samples/boards/pine64_pinetime has the same functionality as the more
general samples/basic/button application with the board specific
driver `pinetime-key-out`.

Signed-off-by: Casper Meijn <casper@meijn.net>

galak reviewed (dismissed):
Is there some doc about the pinetime to understand better what is happening with the "enable-gpio".  I'm not keen on changing the binding of "gpio-keys" in this way.

caspermeijn commented:
> Is there some doc about the pinetime to understand better what is happening with the "enable-gpio". I'm not keen on changing the binding of "gpio-keys" in this way.

Documentation is a bit scarce. https://wiki.pine64.org/wiki/PineTime#Button says:

> The button on the side of the PineTime is disabled by default. To enable it, drive the button out pin (P0.15) high.
>
> While enabled, the button in pin (P0.13) will be high when the button is pressed, and low when it is not pressed. 

And the schematics is available: https://files.pine64.org/doc/PineTime/PineTime%20Schematic-V1.0a-20191103.pdf

Basically 1 gpio provides the power and 1 gpio receives the power. So the first must be an active output, else the second will never detect any input. 

To me this feels similar to `regulator-fixed::enable-gpio` or `semtech,sx127x-base::pa-boost-enable-gpios`.

Glad to hear what would be a better way to document this relationship in the DTS.

galak commented:
> And the schematics is available: https://files.pine64.org/doc/PineTime/PineTime%20Schematic-V1.0a-20191103.pdf
> 
> Basically 1 gpio provides the power and 1 gpio receives the power. So the first must be an active output, else the second will never detect any input.
> 
> To me this feels similar to `regulator-fixed::enable-gpio` or `semtech,sx127x-base::pa-boost-enable-gpios`.
> 
> Glad to hear what would be a better way to document this relationship in the DTS.

What pin(s) am I looking for in the schematic?

galak reviewed (commented):
Is there any concern with just setting GPIO15 in the board code at boot?

caspermeijn commented:
> What pin(s) am I looking for in the schematic?
On the NRF52 it is P0.15 for the output en P0.13 for the input. 

The button is J6. It is connected via P15_KEY to the output en P14_KEY_IN to the input

> Is there any concern with just setting GPIO15 in the board code at boot?
That would fix the button sample. I was thinking that the hardware relationship should be reflected in the DTS, so that a future driver can take advantage of it. But I can think of a better strategy then using the interrupt.

I think setting GPIO15 during boot is the best option. Is it possible to define an always-on gpio in DTS to avoid board-specific code?

galak commented:
> I think setting GPIO15 during boot is the best option. Is it possible to define an always-on gpio in DTS to avoid board-specific code?

No, we don't have anything like that.  We'll need some board code, but it could be driven by DTS if you want.

caspermeijn committed:
boards: arm: pinetime: Fix key definition

The PineTime has only 1 key, which is in series between two gpio. The
`key_out` gpio should to be set high. Then you can read the button
state from the `key_in` gpio.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn commented:
I changed to mostly board specific code. I had to create a dts binding to make it work.

caspermeijn commented:
Now all of the needed code for enabling the key_out pin is board-specific.

caspermeijn commented:
@galak I believe that I made all requested changes. Is there anything I can do to make progress on this PR?

cfriedt commented:
@caspermeijn - please rebase and fix conflicts
@galak - PTAL

caspermeijn commented:
@cfriedt @galak rebased and fixed conflicts

mbolivar-nordic commented on boards/arm/pinetime_devkit0/key_out.c:
```diff
@@ -0,0 +1,42 @@
+
+LOG_MODULE_REGISTER(pine64_pinetime_key_out);
+
+#define KEY_OUT_NODE DT_INST(0, pine64_pinetime_key_out)
+
+#if DT_NODE_HAS_STATUS(KEY_OUT_NODE, okay)
+
+static const struct gpio_dt_spec key_out = GPIO_DT_SPEC_GET_OR(KEY_OUT_NODE, gpios, { 0 });
```
Hard-coding DT_INST numbers like this is generally a bad practice that I am trying to avoid doing more in mainline zephyr, and in this case it looks a bit suspicious.

I say "mostly" because this is a file in the board directory, so you know for sure that the key_out node exists. So my questions are:

- Why not just use DT_PATH(key_out) to get a node identifier, since you know the node exists, and where it is?
- Why not use GPIO_DT_SPEC_GET? That won't fail, because you know the node has the right property

You can hard-code both the node and its property here in board-specific code, is basically what I am saying.

caspermeijn replied:
>     * Why not just use DT_PATH(key_out) to get a node identifier, since you know the node exists, and where it is?

Because it didn't know this is bad practice. To me `DT_INST` feels flexible, as you are not dependent on the name of the node. I changed it to the suggested DT_PATH.

>     * Why not use GPIO_DT_SPEC_GET? That won't fail, because you know the node has the right property

You are right. If the node exists, then I know it has the right property. I removed the `_OR` and the if-statement.

mbolivar-nordic replied:
> `DT_INST` feels flexible, as you are not dependent on the name of the node

Right, but you also have no idea which node you are getting if there is more than one. Often, people don't realize that instance numbers have few restrictions on their order. That's why I generally think hard-coded instance numbers are a bad idea in general, as they often lead to mistakes that just happen to work for the one developer's build when they were writing the code that day. There are helpers like `DT_COMPAT_GET_ANY_STATUS_OKAY` (as well as higher level related helpers like `DEVICE_DT_GET_ANY` and `DEVICE_DT_GET_ONE`) for situations where you really don't care which one you are getting, like in samples.

Hopefully that makes sense, thanks for making the change.

mbolivar-nordic commented on boards/arm/pinetime_devkit0/key_out.c:
```diff
@@ -0,0 +1,40 @@
+
+#define KEY_OUT_NODE DT_PATH(key_out)
+
+#if DT_NODE_HAS_STATUS(KEY_OUT_NODE, okay)
```
Do you really intend to allow users to disable the node? In my previous review, I was suggesting removing this too, or at least that's what I meant to suggest. If you really want to let people turn off this node, this is fine. Just confirming.

caspermeijn replied:
Well, this driver is needed to make the sample applications work. If an application wants to do something smart with the button (only enable during polling for example), then the application can disable this node and create it's own driver.

I don't know whether that is a realistic use-case, but that was my thinking.

mbolivar-nordic reviewed (approved):
One question that I would ideally like an answer to before this is merged, but it's not a big deal and can always be fixed up later if necessary, so it's not a blocking review.
